erc20/src/lib.rs

```diff
@@ -29,16 +30,42 @@ impl ERC20 {
 
         self.balance.write(from, from_balance - value);
         self.balance.write(to, to_balance + value);
+
+        emit!("Transfer", idx from, idx to, value);
```

Comment:
Isn't it the case that in Solidity one defines which fields are indexed when _declaring_ the event, not when using it? What will happen if another `emit!` call doesn't do `idx from`? Or change the (type) signature of the `Transfer` event entirely.

Reply:
Yes, struct makes more sense. 

As for emitting there a couple of options:
1. What you did above. 
2. Construct a `Transfer` instance and pass that to the `emit!`
3. Have `event` generate a `transfer!` macro (named after the struct). 

Although one might argue that 2&3 are just sugar in top of 1.

Reply:
I've just uploaded the update, would the structure be something like this?

Now I need to test everything byte by byte to ensure that everything matches 100% of the time

Reply:
That's awesome @scab24 , thanks for doing this. 

Please, don't treat everything I write as a _command_ :) I really appreciate jumping on this immediately but I don't want to add you work unnecessarily. Let's see what @gakonst and @leonardoalt think as well.

Reply:
@deuszx 
Thank you for your comments, I'm glad to do it, it's a really interesting project 😀

Whenever I have some free time at work, I spend a bit on it

Reply:
I actually like the struct idea. What I don't necessarily like 100% is the `emit!` macro, because the type isn't being used by the user. Currently it's being used like this:
```rust
emit!(Transfer, from, to, value);
```
In this version, the type is only coupled with the arguments because of the macro. In this case we don't really need the struct to exist, the macro could do everything.

```rust
log::emit(Transfer::new(from, to, value));
```
An alternative to that would be the above, where the user gives the full instance, and uses a function instead of a macro. The function could rely on a trait "EncodeLog" (for example), which is implemented for the type when deriving `Event`.

@scab24 let's all agree on a solution here before you implement anything

@gakonst wdyt?

Reply:
I understand, so how would you like to approach it? I can propose it and have these versions available

@leonardoalt 
@deuszx 
@gakonst

Reply:
> I don't think that's true. We need to know the "name" of the event to serialize it properly.

@deuszx true but I guess that's available in the derive macro?

Reply:
@leonardoalt maybe a misunderstanding on my part. I thought you meant sth like this:
```rust
log::emit(from, to, value);
```
but maybe you meant?
```rust
log::emit(Transfer::new(from, to, value));
```

Reply:
yes, that's also what I wrote above haha. It's also similar to what @gakonst said here: https://github.com/r55-eth/r55/pull/14#discussion_r1847265274

@scab24 do you think that works?

Reply:
Then how does that _not_ need a struct? You need a struct if you want to initialize it. It also helps when generating the ABI of the contract I suppose? 

This is Rust, so I'd opt for sticking to "Rust syntax and rules". After all, this project wants to make it easy for _"Rust developers to write Solidity contracts"_ so inventing new syntax seems like a bad idea (this is about @gakonst 's idea from the other thread).

Reply:
Sure, I agree with that too. I didn't say we don't need a struct, I said I like the struct idea and what I don't like 100% is the macro.

Reply:
Agreed -- let's go with this.